nautobot/extras/filters.py

```diff
@@ -418,55 +424,152 @@ class Meta:
+        }
+
         custom_fields = CustomField.objects.filter(
             content_types=ContentType.objects.get_for_model(self._meta.model)
         ).exclude(filter_logic=CustomFieldFilterLogicChoices.FILTER_DISABLED)
         for cf in custom_fields:
-            self.filters["cf_{}".format(cf.name)] = CustomFieldFilter(field_name=cf.name, custom_field=cf)
+            # Determine filter class for this CustomField type, default to CustomFieldBaseFilter
+            new_filter_name = f"cf_{cf.name}"
+            if cf.type in custom_field_filter_classes:
+                filter_class = custom_field_filter_classes[cf.type]
+            else:
+                filter_class = CustomFieldBaseFilter
+            new_filter_field = filter_class(field_name=cf.name, custom_field=cf)
+
+            # Create base filter (cf_customfieldname)
+            self.filters[new_filter_name] = new_filter_field
+
+            # Create extra lookup expression filters (cf_customfieldname__lookup_expr)
+            self.filters.update(
+                self._generate_custom_field_lookup_expression_filters(
+                    filter_name=new_filter_name, filter_field=new_filter_field
+                )
+            )
+
+    @staticmethod
+    def _get_custom_field_filter_lookup_dict(existing_filter):
+        # Choose the lookup expression map based on the filter type
+        if isinstance(existing_filter, (CustomFieldNumberFilter, CustomFieldDateFilter)):
+            lookup_map = FILTER_NUMERIC_BASED_LOOKUP_MAP
+        elif isinstance(
+            existing_filter, (CustomFieldBooleanFilter, CustomFieldJSONFilter, CustomFieldMultiSelectFilter)
+        ):
+            lookup_map = None
+        else:
+            lookup_map = FILTER_CHAR_BASED_LOOKUP_MAP
+
+        return lookup_map
+
+    # TODO 2.0: Transition CustomField filters to nautobot.utilities.filters.MultiValue* filters and
+    # leverage BaseFilterSet to add dynamic lookup expression filters. Remove CustomField.filter_logic field
+    @classmethod
+    def _generate_custom_field_lookup_expression_filters(cls, filter_name, filter_field):
+        """
+        For specific filter types, new filters are created based on defined lookup expressions in
+        the form `<field_name>__<lookup_expr>`. Copied from nautobot.utilities.filters.BaseFilterSet
+        and updated to work with custom fields.
+        """
+        magic_filters = {}
+        # Choose the lookup expression map based on the filter type
+        lookup_map = cls._get_custom_field_filter_lookup_dict(filter_field)
+        if lookup_map is None:
+            # Do not augment this filter type with more lookup expressions
+            return magic_filters
+
+        # Create new filters for each lookup expression in the map
+        for lookup_name, lookup_expr in lookup_map.items():
+            new_filter_name = f"{filter_name}__{lookup_name}"
+            new_filter = type(filter_field)(
```

Comment:
Just making sure I understand this right - the new extended filters will also be single-value, not multi-value filters? Is there any reasonably straightforward way we could make the new filters at least be multi-value so that it's one less breaking filter change we'll have to make in the future?

Reply:
I think the multi-value filters should work for the new extended filters. I didn't do that yet because I thought it may be less confusing for all filters on custom fields to be single value in 1.x and multi-value in 2.x. But if that's an acceptable trade-off then I'll see if the multi-value filters will work here.